src/data/nat/prime_factorization.lean

```diff
@@ -0,0 +1,78 @@
+/-
+Copyright (c) 2021 Stuart Presnell. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Stuart Presnell
+-/
+import data.nat.prime
+import data.nat.mul_ind
+
+/-!
+# Prime factorizations
+
+ `n.prime_factorization` is the finitely supported function `ℕ →₀ ℕ`
+ mapping each prime factor of `n` to its multiplicity in `n`.  For example, since 2000 = 2^4 * 5^3,
+  * `prime_factorization 2000 2` is 4
+  * `prime_factorization 2000 5` is 3
+  * `prime_factorization 2000 k` is 0 for all other `k : ℕ`.
+
+-/
+
+open nat finset list finsupp
+open_locale big_operators
+
+namespace nat
+
+/-- `n.prime_factorization` is the finitely supported function `ℕ →₀ ℕ`
+ mapping each prime factor of `n` to its multiplicity in `n`. -/
+noncomputable def prime_factorization (n : ℕ) : ℕ →₀ ℕ := (n.factors : multiset ℕ).to_finsupp
+
+lemma prime_factorization_count {n p : ℕ} : n.prime_factorization p = list.count p n.factors :=
+by { unfold prime_factorization, simp }
+
+/-- Every positive natural number has a unique prime factorization -/
+lemma prime_factorization_eq_iff {a b : ℕ} (ha : 0 < a) (hb : 0 < b) :
+  a.prime_factorization = b.prime_factorization ↔ a = b :=
+begin
+  split,
+  { intros h,
+    refine eq_of_count_factors_eq ha hb (λ p, _),
+    simp only [←prime_factorization_count, h] },
+  { intros h, rw h },
+end
+
+lemma prime_factorization_zero : prime_factorization 0 = 0  :=
+by { unfold prime_factorization, rw factors_zero, simp }
```

Comment:
`simp` will unfold, and stuff like `factors_zero` are `simp` lemmas; you can replace a lot of these by blocks with just `by simp [prime_factorization]`.

Reply:
Thanks!